Add DataModel 2 spec with an experimental syntax

matthewmueller opened the pull request.

schickling commented:
A couple of points to discuss/rethink:

- [ ] How to reduce the learning curve / make the datamodel syntax feel more familiar
  - Ideas:
    - Use `field: Type` instead of `field Type`
    - Use `@attribute` instead of `attribute()`
- [ ] Comma/multi-line behavior seems error-prone
- [ ] Relations: `Model@field` doesn't feel right
- [ ] `primary() cuid()` seems like a very common use case. Can we make this more compact/convenient?
- [ ] New feature: Ability to split up the datamodel into multiple files and have some sort of explicit/implicit import mechanism
- [ ] Is "datamodel" still the right name as it now contains configuration options? Is it the right decision to merge both together? (This is a big decision!)
- [ ] Spec out endpoint/secret configuration options

Generally we should also prepare a couple of real-world examples to get a better feeling for the syntax.

mavilein commented:
Answer to Open Questions:
> 4. Can back-relations have a different nullability than the forward relations?
Yes this is possible.

> 5. string or text?
My personal association is that string is something that has a max length and can be efficiently indexed. I associate text with larger paylods.
One example: I would intuitively choose `string` for `Post.title` but `text` for `Post.content`.

> 9. Can we rename datamodel to schema?
Please no. We just decided to switch from the term schema to datamodel in the backend :joy:

mavilein commented:
**Requests**
- [ ] Can you add an example of how indexes would look like in that syntax? There’s an example for a combined unique index but not for a normal index that just exists to speed up things

mavilein commented:
email          postgres.Citext  unique() postgres.Like(“.%.com”)
Specifying a field like this makes it harder to read the Datamodel imo. When reading the Datamodel i am often just interested in the shape of data available for me in the client. In that case i need to map `postgres.Citext` to `String` in my head. I would therefore like to always see the common type such a special type is mapped to.
> email          String unique() postgres.Like(“.%.com”) postgres.Citext
This also makes conversion to other databases much easier because you can simply remove all instances of `postgres.xxx` in your Datamodel.

ejoebstl commented:
Things that I like: 
* It takes multi-database schemas into account. Relations might exist across shared databases.
* Anonymous embedded types. That's a very idiomatic way of defining relations. We might want to discuss allowing that even for non-embedded types.
* Explicit declaration of the reference column in relations, `Writer@id`. This also solves ambiguous relations. Id should be the default though.
* Join tables. I think the possibility to add fields to a join table is amazing.

Things I'm unsure about:
* Forcing join tables. I would suggest to make them optional to keep make getting started with prisma more simple.
* Usage of `@` to declare reference columns. Maybe we can just use a `.`, like when accessing the field of an object?

Thing that I don't like: 
* Mixing database (endpoint) configuration with the datamodel. That should be two entirely different things. You might, for example, want to use the same datamodel for three databases (dev, staging, production).
* The  sql-like notation of `id     int   primary() serial()`. I find it hard to read, compared to the classic notation.
* Using `default(...)` instead of ` = `. It adds complexity to the syntax, but it is easier to read.

sorenbs commented:
# Answer to Johannes's comments

> * [ ]  How to reduce the learning curve / make the datamodel syntax feel more familiar
>   
>   * Ideas:
>     
>     * Use `field: Type` instead of `field Type`
>     * Use `@attribute` instead of `attribute()`

It should not be a goal to make it familiar to people who know GraphQL SDL. Prisma Datamodel should feel accessible to as many development (js, python, java etc) and database (Postgres, Mongo, Cassandra etc,) communities as possible. Common syntax within these communities vary widely. Our best shot at a common denominator is probably something JSON-like, but I don't think we should restrain ourselves too much.

> * [ ]  Comma/multi-line behavior seems error-prone

Agree. I appreciate that it gives us other freedoms, but it will be unexpected behavior to most developers. I we decide to use a special construct to support multi-line, we should try to find something more obvious or common.
There are other ways to introduce multi-line support without requiring a special construct like this, but they introduce restrictions in the grammar. 

> * [ ]  Relations: `Model@field` doesn't feel right

I really like this suggestion as a way to break out of default behavior. The default should still be to reference the `@id` (or whatever syntax we choose for it) field (Primary Key in relational databases and `_id` in Mongo)

> * [ ]  `primary() cuid()` seems like a very common use case. Can we make this more compact/convenient?

We could introduce the convention that a field named `id` will be chosen as the `@id` field unless something else is specified.

I would like to see if we can find a syntax that makes it clear that `cuid()` is a generator as described in the original Datamodel 2 spec. Maybe something like this: `default(cuid())` or `= cuid()`

> * [ ]  New feature: Ability to split up the datamodel into multiple files and have some sort of explicit/implicit import mechanism

Yes, we need this. Especially important if we move configuration into the same format.

> * [ ]  Terminology: Is "datamodel" still the right name as it now contains configuration options? Is it the right decision to merge both together? (This is a big decision!)

I generally like merging config and schema specification as long as we have a good system for splitting into multiple files. We should consider some convention around managing multiple environments.

I still like the term datamodel.

> * [ ]  Terminology: Is `model` the right term in the context of entities etc?

I don't think we have anything better atm.

sorenbs commented:
# Answer to open questions

> **1. Lowercase primitives, uppercase indentifiers?**
> 
> - If primitives are considered special and cannot be overridden
>   then I think we should have special syntax for them. If they are
>   simply types that are booted up at start (GraphQL), they should
>   be treated like every other type.

I think primitive types are special. They are a carefully selected set of types that should be implemented by most connectors. They can not be overridden, but type specification can be used to influence their implementation in a particular field. For example, I don't like this syntax as it is not portable, and there is no obvious way to make it portable:

```groovy
email          postgres.Citext  unique()
```

Instead I would like to rely on a generic type specification syntax like this

```groovy
email          string  unique() typeSpecification(postgres.Citext)
```

This way a connector can pick up the type specification if it understands it and fall back to the more generic primitive type (string in this case) if it doesn't. The concrete syntax for this is still open question.


> **2. Do we want back-relations to be optional?**
> 
> - My typical stance is to enforce good practices (e.g. prettier),
>   and provide one way to do it. We have some options with back-relations though:
> 
>   1. Implied back-relation when not provided (affects client API)
>   2. No back-relation when not provided (affects client API)
>   3. Build-time (`prisma generate`) error when no back-relation provided

1. is reasonable, and I think we have found that we can handle this in query engine without too much trouble
2. doesn't make sense
3. could be an option if we implement proper assistance in VS code plugin. This would simplify the query engine, but I am not convinced we need this.

> **3. Should we enforce link tables?**
> 
> - Link tables are not usually needed right away, but are often good practice
>   since you often want to attach metadata to that relation later on
>   (e.g. `can_edit bool`). Some options:
> 
>   1. We could make them optional at first, but create a table in the background (we'd need to do this anyway), but then when they specify the table and migrate, we'll be aware that this implicit join table became explicit in the datamodel
> 
>   2. Enforce the link table at build-time when we run `prisma generate`. A bit simpler to implement and less magic, at the expense of cluttering up your datamodel file and forcing you to think more about your data layout earlier on (might not be a bad thing).

First, Prisma 1 used Link tables for all relations (1:1, 1:m, n:m). This was bad for performance, and we now only use link tables for n:m relations. Marcus convinced me that having a default behavior for implicit link tables for m:n relations is a good idea as it provides two benefits: 1) simpler to get started and 2) portability between different databases that implement m:n relations in different ways.
I think this one is settled :-)

> **4. Can back-relations have a different nullability than the forward relations?**

Yes, as Marcus said

> **5. `string` or `text`?**
> 
> 🙃
> 
> - String is more familiar to programmers.
> - Text is more familiar to English speakers
> - Text is shorter.

I generally think we should pick "developer terms". Marcus suggest we could have both: 

```
My personal association is that string is something that has a max length and can be efficiently indexed. I associate text with larger paylods.
One example: I would intuitively choose string for Post.title but text for Post.content.
```

The problem I see with this is that there is no such distinction that is common across databases. That's why I suggest we have only one string type and let developers implement type specification for their particular database if they need it for performance. 
 
> **6. Should we support Mongo's many-to-many relationship in relational databases?**
 
I think this is a very advanced use case that we can ignore for now. Generally, separating read- and write-path is a better approach to this as discussed in the book Data-Intensive Applications.

> **7. Should we have named arguments?**
> 
> I generally like named arguments more, but with this new syntax it feels
> like named arguments could generally be different attributes:
> 
> ```
> db(name: "users")
> ```
> 
> to
> 
> ```
> db("users") or name("users") or db.name("users")
> ```
> 
> In the cases where it does matter, we could maybe get assistance from
> VSCode on this one. Here's what Android Studio does for Kotlin:
> 
> ![named params](https://cldup.com/bXwmV_70KW.png)

Let's map out all the syntax and try both approaches :-)

I generally dislike the Kotlin style IDE magic. Let's rather design the syntax so it is not needed.

> **8. Should `id`, `created_at`, `updated_at` be special types that the database adds?**
> 
> The DM2 proposal proposes:
> 
> ```groovy
> model User {
>   id: ID @id
> }
> ```
> 
> The reason I avoided this is because it's not always clear (at least in SQL databases)
> whether you want your ID to be an `int`, a `text`, or a `postgres.UUID`. I could
> definitely see value in higher-level types where Prisma chooses the datatype based
> on the most common cases. We'd just need to have a way to override them.
> 
> Same could go for `created_at` or `updated_at`. Where they bring a type and
> some default functionality.
> 
> ```groovy
> model User {
>   id         id
>   createdAt  created_at
>   updatedAt  updated_at
> }
> ```
> 
> One possible solution is to introduce a prisma namespace for high-level data types:
> 
> ```groovy
> model User {
>   id         prisma.ID
>   createdAt  prisma.CreatedAt
>   updatedAt  prisma.UpdatedAt
> }
> ```

A few things here:

I propose we get rid of the ID type. Instead we should propose the best practice to declare a custom primitive type that encapsulates how this particular application want the ID field to behave:

```groovy
type ID = String @cuid @postgres.type(varchar(25))
```

I propose we remove `CreatedAt` `UpdatedAt`. The first is easily covered by Dynamic Generators:

```groovy
model User {
  createdAt: DateTime = now()
}
```

`UpdatedAt` is different as we would need a trigger concept to deal with this in a more generic way. I propose that we don't include it initially and think about how to introduce a trigger system in the future. This probably goes together with thoughts on StratoQL and Expression Generators.

> **9. Can we rename datamodel to schema?**
> 
> I learned that the historical context for datamodel was that we needed something different than schema.graphql.
> 
> I think we should revisit this because `schema.prisma` is much shorter and sounds nicer.

I still like datamodel. I think we will understand the role of the datamodel much better in a years time, and suggest we delay any renaming discussion til then.

sorenbs commented:
# Answer to Marcus's general comments

### Indexes

I like the idea of moving everything that is not field-specific into dedicated blocks. We should map out if we just want a single `meta` block or should rather have many purpose specific blocks:

```groovy
model User {
  dataSource = {
    // use specific table name
    name = "tbl_user"
  }
  generator = {
    // adjust the plural form used by client generators 
    plural = "people"
  }
  index = {
    unique(email, name)             alias("email_name_index")
  }
}
```

I am not sure indexes should be a primitive supported directly by Prisma Core. [Datamodel V2 proposes that Indexes are entirely managed by connectors](https://github.com/prisma/rfcs/blob/datamodel/text/0000-datamodel.md#indexes), but connectors are allowed to into consideration two Prisma Core concepts when resolving the set of indexes:

Indexes are storage engine specific and mostly relevant for performance configuration. Therefore, index configuration is provided by connectors.

> There are two exceptions where indexes intersect with data modelling:
> 
> - Most storage engines implement the _unique constraint_ as an index. Unique constraint is provided by Prisma core, and the connector can choose to create only a single index if both a index and unique constraint is present on a single field in the datamodel.
> - The concept of a _primary key_ is provided by Prisma Core (`@id`). Many storage engines implement the primary key using a special index (sometimes called clustered index or primary index) that organises the data on disk by that field, even if no index is specified separately. These connectors will allow you to configure the index used for the primary key separately using the connector specific index configuration.

A further observation: If we decide to introduce purpose-specific blocks, we probably want to find a way to scope them to a specific datasource:

```groovy
model User {
  dataSource = {
    postgres = {
      // use specific table name
      name = "tbl_user"
    }
  }
}
```

```groovy
model User {
  postgres = {
    datasource = {
      // use specific table name
      name = "tbl_user"
    }
  }
}
```

```groovy
model User {
  dataSource@postgres = {
    // use specific table name
    name = "tbl_user"
  }
}
```

### Config

I like the simplicity of having only a single format and only a single file for simple cases. Your concerns are valid, but I think we can provide a good experience by having a solid system for combining multiple datamodel files. We need to find the right mental model for this, so it si generic enough to support different use cases and simple enough that it feels good to use.

### Relation tables

I don't want us to introduce a concept of through tables. I believe this introduces complexity for little gain. The [Explicit join model](https://github.com/prisma/rfcs/blob/datamodel/text/0000-datamodel.md#explicit-join-model) section lays out how I think about this.

In the future we could decide to introduce a new concept [Edge Relation](https://github.com/prisma/rfcs/blob/datamodel/text/0000-datamodel.md#edge-relation) that fits the Graph database model and simplifies the generated client API similar to what you describe as a benefit for Active Record

### Relations referencing fields

I like this idea as well. I think we should default to the primary key (rather than a field that happen to be called `id`)

mavilein commented:
I have reread the proposal now and i discovered one missing piece: How will the user specify relation settings? I am asking for the equivalent of:
```
type Blog {
  posts: [Post] @relation(name: "BlogToPosts", link: INLINE, onDelete: CASCADE)
}
```

mavilein commented:
I am still a bit unsure whether this proposal makes the datamodel a more dense and therefore harder to read. I think we also have to take into account that people might not want to invest the time to vertically align everything. For users of vs code we will be able to provide a plugin to do that but we should also have non vs code users in mind. 
Therefore i made a comparison of one datamodel in three different styles:

1. With vertical alignment:
```
model User {
  meta = {
    // adjust name from "users" convention
    name = "people"
  }

  // model fields
  id             int              primary() serial
  email          postgres.Citext  unique() postgres.Like(".%.com")
  name           text?            check(name > 2)
  role           Role
  profile        Profile?         alias("my_profile")
  createdAt      datetime         default(now())
  updatedAt      datetime         onChange(now())

  weight         Numeric          alias("my_weight")
  posts          Post[]

  // composite indexes
  unique(email, name)             alias("email_name_index")
}
```

2. Without vertical alignment:
```
model User {
  meta = {
    // adjust name from "users" convention
    name = "people"
  }

  // model fields
  id int primary() serial
  email postgres.Citext unique() postgres.Like(".%.com")
  name text? check(name > 2)
  role Role
  profile Profile? alias("my_profile")
  createdAt datetime default(now())
  updatedAt datetime onChange(now())

  weight Numeric alias("my_weight")
  posts Post[]

  // composite indexes
  unique(email, name) alias("email_name_index")
}
```

3. Without vertical alignment based on the first SDL based approach:
```
@db(name:"people")
@indexes(value: [
  { fields: [.email, .name] name: "email_name_index" }
])
model User {
    id: int @id(strategy: SEQUENCE)
    email: string @unique @postgres.citext @postgres.like(".%.com")
    name: string? 
    role: Role
    profile: Profile? @db(name:"my_profile")
    createdAt: datetime = now()
    updatedAt: datetime = onChange(now())
    weight: Numeric @db(name:"my_weight")
    posts: [Post]
}
```

sorenbs commented:
# Answer to Emanuel's comments

> Things that I like:
> 
> * It takes multi-database schemas into account. Relations might exist across shared databases.

Yes - great!

> * Anonymous embedded types. That's a very idiomatic way of defining relations. We might want to discuss allowing that even for non-embedded types.

+1 for Anonymous embedded types. I don't think we should extend this to non-embedded types as the main difference from a data-modelling perspective is that models can be directly addressed (queried by id for example), and and embedded types cannot. We could allow configuring the storage mechanism for embedded types, such that they can either be stored inline in the same record (as now) or in a different table/document, but still behave like embedded types. I don't think we need to map this out now though.

> * Explicit declaration of the reference column in relations, `Writer@id`. This also solves ambiguous relations. Id should be the default though.

Agree - this is awesome. The default should be the primary key field, not a field named `id`.

> * Join tables. I think the possibility to add fields to a join table is amazing.

Yes. See my comments above.

> Things I'm unsure about:
> 
> * Forcing join tables. I would suggest to make them optional to keep make getting started with prisma more simple.

Agree

> * Usage of `@` to declare reference columns. Maybe we can just use a `.`, like when accessing the field of an object?

Not sure about the syntax either, but I like the concept

> * Enums with a type. That's not supported by all databases (MySQL cannot do that).
> 
> Thing that I don't like:
> 
> * Mixing database (endpoint) configuration with the datamodel. That should be two entirely different things. You might, for example, want to use the same datamodel for three databases (dev, staging, production).

I actually think I like this. We have to have a very convenient way to combine datamodel files or import declarations. I'm sure we can figure this out.

> * The  sql-like notation of `id     int   primary() serial()`. I find it hard to read, compared to the classic notation.

I agree with this.

> * Using `default(...)` instead of `=`. `=` adds complexity to the syntax, but it is easier to read.

No strong opinion on this

> * Unicode (emoji) support: Maybe in string literals, but I see a thousand cases where database systems just explodes if a column or table gets named 😋.

Good point. We want to support literals for sure.

sorenbs commented:
@mavilein - we might want to force (or at least enable) pretty printing through the CLI or any other workflow where we can hook in. I think it is a common expectation to have pretty printing (at least in js/ts, rust, c#, scala language communities). So I think the non-pretty printed version being less readable is a non-issue.

matthewmueller commented:
Hey guys thanks so much for providing all this feedback. I spent yesterday and this morning incorporating it into a revised spec.

As a starting point, it may by useful to start looking at the open questions, which I tried to bring in the feedback and address the common concerns: https://github.com/prisma/rfcs/blob/0001-datamodel-2/text/0001-datamodel-2.md#can-we-make-syntax-more-familiar

For an overview of what's changed, it's probably useful to skim over the full diff: https://github.com/prisma/rfcs/pull/16/commits/a742ce56abe582e25b00c0f8104b545a352e26c2

I hope I was able to answer or address all of your points but if I didn't adequately address your concerns please let me know.

matthewmueller commented:
Hey guys, incorporated the previous discussions with @mavilein and @sorenbs. The syntax across the spec isn't up-to-date, but I pushed answers to our open questions and updated the remaining open questions:

- Diff: https://github.com/prisma/rfcs/pull/16/commits/b231cf6e3f43c48f23a37e4ac60fd8d7efa42c5e
- Q&A from discussion: https://github.com/prisma/rfcs/blob/0001-datamodel-2/text/0001-datamodel-2.md#lowercase-primitives-uppercase-indentifiers
- Remaining Open Questions: https://github.com/prisma/rfcs/blob/0001-datamodel-2/text/0001-datamodel-2.md#open-questions
- Upcoming Syntax adjustment:  https://github.com/prisma/rfcs/blob/0001-datamodel-2/text/0001-datamodel-2.md#answer-9

ejoebstl commented:
1) Syntax Changes: When we start to use `@` for attributes, you can also throw out the `{` and `}` and recognize type by the `model` keyword. Not saying that we should though, as we can no longer have fields called `model` then. 
2) I still have stomach ache about dropping the `:` and requiring indentation.
   * It's a burden to maintain without strong editor support. Not everybody out there uses VSCode. 
   * Git commits will be polluted when adding a field and reformatting. 
   * A single very long field in a type makes everything hard to read (don't forget that a BIG use case of prisma is legacy DBs!). 
3) Alternative suggestion: What about classic C-style syntax without indentation: `string firstName @id @bla`?
4) Re rendering the datamodel against real world examples: We should take the full complexity into account there, against the nastiest legacy samples we have, including indices, sequences, different kind of relations, and maybe even declarations for custom types and triggers.

I had a few more thoughts on the multi-line and type directive stuff, but I will wait until the spec is updated.

matthewmueller commented:
> Syntax Changes: When we start to use @ for attributes, you can also throw out the { and } and recognize a type declaration by the model keyword. Not saying that we should though, as we can no longer have fields called model then.

This is an interesting idea! I think this would conflict with model-level attributes under the new spec though (composite unique fields):

```groovy
model User
  id                string @primary
  email          string
  first_name  string
@unique(email, first_name) // this attribute would be applied to the first name, not the model
```

> I still have stomach ache about dropping the : and requiring indentation.
It's a burden to maintain without strong editor support. Not everybody out there uses VSCode.
Git commits will be polluted when adding a field and reformatting. A single very long field in a type makes everything hard to read (don't forget that a BIG use case of prisma is legacy DBs!).

I feel you. I'd just like to be conservative with our syntax since we might need that later. I think you'll get used to this very quickly.

Regarding formatting, it's important to clarify that indentation is not required. This is valid:

```
model User {
  id string @primary
  email string
}
```

It'll look nicer if it's formatted though. As I'm sure you know already, current industry practice is to ship standalone CLI tools (e.g. `gofmt`, `prettier`) which allow you to do stuff like this `gofmt < main.go > main.go`. I think it's reasonable to assume most modern editors can run a command on save. And if you're using notepad, then you can use the CLI tool standalone ☺️

Regarding big types, we're currently thinking of the type field as only supporting primitive types (e.g. `string`, `int`, etc.) and types you define in your datamodel (e.g. `User`). You can "upgrade" those types for a specific database with a type spec. Something like this:

```groovy
model User {
  id  string  @as(postgres.UUID) @as(mongo.ObjectID)
}
```

> Alternative suggestion: What about classic C-style syntax without indentation: string firstName @id @bla?

Interesting! I hadn't thought of that though I did wonder about:

```groovy
Post model {
  id     string  @primary
  title  string
}

Role enum {
  PUBLISHED
  DRAFT
}
```

The reason Go broke with tradition and went with `name type` is that they thought that the name was more important than the type. I think that's also true here, but I'd be curious what other people think.

> Re rendering the datamodel against real world examples: We should take the full complexity into account there, against the nastiest legacy samples we have, including indices, sequences, different kind of relations, and maybe even declarations for custom types and triggers.

100% agreed. I see this as a huge limitation we have right now. Not enough real-world samples.

matthewmueller commented:
FYI, I've updated the syntax and the spec based on our current thinking. A couple highlights:

**Type Specifications / Upgrades**

I'm thinking of using `@as(...)` attribute

```groovy
model User {
  id  string  @as(postgres.UUID) @as(mongo.ObjectID)
}
```

**Blocks inside blocks**

Embedded embeds are workable in the current model but I think may lead to confusion since adding a "=" turns them into an Object. I think we should enforce that blocks have 2 (or possibly more) namespaces.

From:

```groovy
// named embed (reusable)
embed Photo {
  id   text    @as(mgo2.ObjectID)
  url  string

  // anonymous embed (optional)
  size {
    height int
    width int
  }?

  // anonymous embed
  alternatives {
    height int
    width int
  }[]
}
```

To:

```groovy
// named embed (reusable)
embed Photo {
  id   text    @as(mgo2.ObjectID)
  url  string

  // anonymous embed (optional)
  embed size {
    height int
    width int
  }?

  // anonymous embed
  embed alternatives {
    height int
    width int
  }[]
}
```

It's not as nice, but I'm not sure it's terrible either. This also allows the syntax to have nested blocks. What do you guys think?

**Moved model attributes below model block**

The new syntax will look like this:

```groovy
model Post {
  id     int     @primary
  title  string
  author Author
}
// unique composite
@unique(title, author)
```

The reason to bring the model attributes below is that it adds consistency to the attribute syntax and solves multi-line issues.

```
block-type block-name {
  field-name field-type  field-attributes
}
block-attributes
```

matthewmueller commented:
Feedback from @schickling regarding *Blocks inside Blocks*

> I’d probably switch it around here to `size embed`
> Would need to see a few examples. Still feel like I prefer `size embed`, even if it’s less consistent



To:

```groovy
// named embed (reusable)
embed Photo {
  id   string    @as(mgo2.ObjectID)
  url  string

  // anonymous embed (optional)
  size embed {
    height  int
    width   int
  }?

  // anonymous embed
  alternatives embed {
    height  int
    width   int
  }[]
}
```

Where the AST looks more like this:

```
block-type block-name {
  field-name field-type  field-attributes

  block-name block-type {
    field-name field-type  field-attributes
  }
}
block-attributes
```

mavilein commented:
> I'm starting to think that it might be best to use the low-level field names as the default for datamodel 2.

I think we should follow the mental model of @sorenbs here. He thinks of the Datamodel as the intermediate layer between application and database schema. Therefore i suggest the following rules:
1. The name of a field equals the name of a column and the name of a field in the application schema if not customised.

2. The underlying db name can be customised. This can either be done through a field level directive or a global setting. E.g.:
```groovy
source postgres {
  ...
  name-generators = "snake_case"
}
model Blog {
  meta = {
    from = "postgres"
  }
  id          int
  myTitle     string @db.name("my_title")
  myTitleTwo  string # will be my_title_two due to the name generator
}
```
3. The name of a field in the application can be customised. This can either be done through a field level directive or a global setting. E.g.:
```groovy
client {
    name-generators {
        js   = "camelCase"
        go   = "UpperCamelCase"
        ruby = "snake_case"
    }
}
model Blog {
  id          int
  myTitle     string @client.name("MyTitle")
  myTitle2    string # would be MyTitle2 for the generated Go client due to the name generator
}
```

With this approach your desired behaviour becomes opt in. I understand your reasons but i don't think that this is a default that we should force on our users. The default should be rule 1 that can be easily customised if desired.

-----------------------------------------
> I didn't quite understand the implementation details of why enums would be lowercase. Maybe @marcus and @sorens want to clarify here?

The reason was that an enum maps to a primitive type in the database. And since we are going to lowercase primitve types we should do the same for enums. The main reason to lowercase primitives was that they can be efficiently indexed. This way all indexable fields are simply lowercase.
While i like this reasoning we are starting to compromise more and more on this idea. We now have 2 spots where this reasoning would be weird:
1. Type aliases. E.g. `type MyId = cuid @primary`
2. Enum names. 

Maybe it is just better to go with all uppercase and be done with those awkward questions.

matthewmueller commented:
Feedback from @mavilein 

> I think we should follow the mental model of @sorenbs here. He thinks of the Datamodel as the intermediate layer between application and database schema. Therefore i suggest the following rules:
> The name of a field equals the name of a column and the name of a field in the application schema if not customised.
> 
> The underlying db name can be customised. This can either be done through a field level directive or a global setting. E.g.:

Yes good call. I agree with this 100%. I'll make sure to put these rules in the spec.

> Maybe it is just better to go with all uppercase and be done with those awkward questions.

Yah, I think the efficiently indexable separation will be a bit hard to explain to people.

> I like the new syntax for anonymous embeds a lot since it is more consistent to me with other fields 🙏

See below

---

I've had a chance to formalize the grammar a bit and double-check things. Overall, I think this feels really nice 😬 A couple things came up during this process, I want to highlight:

**Things in Terraform/HCL that we might not need**

These haven't been provided in this spec (or any other spec) and I'm not sure we'll need. Philosophically, we need to decide if we should be able to support any HCL configuration or if Prisma files are merely HCL-inspired. I'm leaning towards HCL-inspired rather than full superset since I don't see any real benefit in being able to parse HCL files for Prisma.

> Source: https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/spec.md#structural-elements

- Conditional = Expression "?" Expression ":" Expression;
- OneLineBlock = Identifier (StringLit|Identifier)* "{" (Identifier "=" Expression)? "}" Newline;
- https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/spec.md#for-expressions
- https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/spec.md#template-expressions
- Splats: https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/spec.md#splat-operators

My preference would be to leave out this syntax for now. Terraform added it because they needed it. We may need it down the line, but I'd like to make our own decisions here.

**Enum ambiguities**

@ejoebstl suggested that we shouldn't have a type on Enums because we want to have the ability to choose the underlying type. It's also present in the previous spec. I've adopted the no type approach in this spec. It looks like this:

```
enum Role {
  USER   // unless explicit, defaults to "USER"
  ADMIN  @default("A")
}
```

However, I ran into an ambiguity in parsing that would require specific knowledge about the `enum` 
keyword to parse, breaking our nice-to-have requirement of no symbol tables: https://github.com/prisma/rfcs/blob/0001-datamodel-2/text/0001-datamodel-2.md#requirements

This is a sign to "stop & think". I don't think encoding special cases like this would be terrible, but I'm not sure enums should be considered a prisma primitive or should be supplied by the database-specific connectors. Another option would be to change the syntax of Enums to break the ambiguity.

**HCL weirdness**

Hashicorp has some inconsistencies around Object and Arrays and newlines. I think these match the user's expectations, but are pretty weird if you're deep in syntax world.

```hcl
// valid object
obj = { a = "a", b = "b" }

// valid array
arr = [1, 2]

// valid object
obj = {
  a = "a"
  b = "b"
}

// invalid array
arr = [
  1
  2
]
```

> More info: https://github.com/hashicorp/hcl/pull/51

**Embeds**

I keep going back and forth on these. @schickling and @mavilein prefer: 

```groovy
embed Photo {
  url string
  sizes embed {
    width int
    height int
  }[]
}
```

Which would make our syntax look like this:

```
blockType blockName "{" Newline
  (FieldName FieldType FieldMap? FieldAttributes* Newline)*
"}" Newline
```

Adding the `FieldMap` optional in here. I think they would also be okay with:

```groovy
embed Photo {
  url string
  sizes {
    width int
    height int
  }[]
}
```

No `embed` syntax. I think I prefer this and it's what's I originally proposed. The only issue I see is this case:

```groovy
model Photos {
  meta = {
    from = mongo
    name = "photos"
  }
  size {
    width  int
    height int
  }[]
  url string
}
```

Do you guys see any issues with readability by having embeds and configuration next to each other? Do you think they'd be easy to confuse?

**Bringin' Back the Colon**

cc/ @schickling 

I realized last night that we could safely support the `:` from a parsing perspective and use something like `prismafmt` to remove them or replace them with `=`

Blocks:

```groovy
model Post {
  title: string
}
```

Gets formatted into this:

```groovy
model Post {
  title  string
}
```

Objects:

```groovy
meta: {
  a: "b"
}
```

Gets formatted into this:

```groovy
meta = {
  a = "b"
}
```

Future syntax like `fieldAlias ":" fieldName ":"? fieldType` would still be fine. 

**Argument Names**

I've been going back and forth on this one a lot. We have a couple options here:

1. All named arguments `@alias(name: "my_weight") @sequence(start:1, step:100)`
2. Optionally named: `@alias("my_weight") @sequence(step:100, start:1)`. Naming allows you to specify the arguments in whatever order you want.
3. No named arguments, VSCode plugin will provide autocomplete signatures.
4. No named arguments, break up optional attributes when not obvious to be multiple 1 argument attributes `@sequence.start(1) @sequence.step(100)`

/cc @sorenbs @mavilein @ejoebstl @schickling

mavilein commented:
> I'm planning on copying Terraform here in spite of inconsistencies, but let me know if you want me to make any adjustments.

Sounds good to me.

> Do you guys see any issues with readability by having embeds and configuration next to each other? Do you think they'd be easy to confuse?

This is why i liked the `embed` key word because those blocks can be distinguished more easily. But these are just my 2 cents. Let's go with a majority decision here.

> realized last night that we could safely support the : from a parsing perspective and use something like prismafmt to remove them or replace them with =

If we remove it anyway through formatting let's just not have it in the first place.

> Argument Names

- Option 3 is a no go for me because not everyone uses VSCode.
- Option 4 introduces more syntax complexity imo. I would not know when i am supposed to use the dot syntax versus the normal one.
- Option 1 is very consistent but would contradict the overall approach you took with the directives so far where they are most singly argument directives. This would necessitate the grouping of related settings. E.g. consider `@relation(name: "MyRelation" onDelete: CASCADE)` in DM 1.1 vs `@relation(name: "MyRelation") onDelete(value: CASCADE)`. This would also require us to add lots of unnecessary argument names a la `value`.
- The direction of Option 2 seems to be the most promising to me. But we need to spec that out a bit further imo. How do we handle arguments in directives with multiple arguments? Your suggestion implies that `@sequence(100,1)` is possible which is not very readable. Therefore i suggest the following rule:
    - Directives with 1 argument may omit the argument name. E.g. `@alias("my_weight")` and `@alias(name:"my_weight")` is allowed. Directives with multiple arguments must always specify all argument names.

matthewmueller commented:
Sounds good, thanks @mavilein. I agree with your above points, and I like your suggestion for the rule on named arguments, however I'd like more feedback on named arguments because I think optionally named may complicate our existing speced out attributes: https://github.com/prisma/rfcs/blob/0001-datamodel-2/text/0001-datamodel-2.md#should-we-have-named-arguments

mavilein commented on text/0001-datamodel-2.md:
````diff
@@ -1120,8 +1120,15 @@ model User {
+
+```
+Blog[]: empty list or non-empty list of blogs (default: [])
+Blog[]?: empty list, non-empty list of blogs or null (default: null)
````
I don't understand this one. When would Prisma ever return null instead of an empty list?

matthewmueller replied:
Hey so, I've based that off the ability to have these 3 states in Postgres and the assumption that someone will have a valid use case for this:

```
create table users (
  id serial primary key,
  scope text[]
);
```

Personally, I haven't had this use case in my database schemas yet, but I have had cases in Go where I want these 3 states. We very may want to consolidate here thouhg.

sorenbs commented:
All code blocks in this comment are snippets from the spec that I am commenting on. It's a bit unwieldy, so I suggest we can go through them together @matthewmueller. 

# Feedback on Basic Example

> The basic example seems to be inconsistent with other sections in the spec, so let's use this list to make the entire spec internally consistent.

```
// type definition
type Numeric @postgres.Numeric(5, 2)
```

I assume `Numeric` is the name of the new type. I think it would be beneficial that this is bound to one of our build in primitive types, so that generators and potentially other tools in the future have a lowest common denominator to fall back to. My suggested syntax is this:

```
// type definition
type Numeric = Decimal @postgres.Numeric(5, 2)
```

---

```
  meta = {
    // adjust name from "users" convention
    name = "people"
  }
```

This is specifying the database name. For a SQL connector, this is saying that the table is called `people`. Could we use a property name that is more descriptive? For example `connectorName`, `dbName` or similar?

---

```
email          string    @unique  @postgres.Like(".%.com") @as(postgres.Citext)
```

`@as` is the syntax for type specification. 

- I am not sure the word `as` is the most descriptive we can find.
- We should also make sure to use the same syntax for type definitions and fields, so maybe we simply temove the `@as` part:

```
email          string    @unique  @postgres.Like(".%.com") @postgres.Citext
```

is now in line with

```
// type definition
type Numeric = Decimal @postgres.Numeric(5, 2)
```

---

```
profile        Profile?  @alias("my_profile")
```

My understanding is that `@alias("my_profile")` influences the client generator. Maybe we should scope the name of this directive: `@generator.alias("my_profile")`.

Note that an alternative is to use the same name in the datamodel and the generator by using the `@dbName` directive:

```
my_profile     Profile?  @dbName("profile")
```

Also note that the left-most field defines the name at the datamodel level while `@dbName("profile")` defines the name of the column in the database.

---

```
createdAt      datetime  @default(now())
```

I would still consider using `=` as the syntax for the default value, but I don't have a strong opinion on this:

```
createdAt      datetime = now()
```

---

```
createdAt      datetime  @default(now())
updatedAt      datetime  @onChange(now())
```

I would suggest that we temporarily keep the `@updatedAt()` attribute as we don't fully understand the `@onChange()` directive yet. I suspect that we will have a fully fledged trigger system eventually, most likely powered by wasm code executed at the Prisma layer.

Suggestion:

```
createdAt      datetime  @default(now())
updatedAt      datetime  @updatedAt()
```

---

```
enum Role {
  USER   enum  // unless explicit, defaults to "USER"
  ADMIN  enum  @default("A")
}
```

I don't think "defaults to" is the right terminology here. Using `=` to specify the underlying primitive value can avoid this awkwardnes:

```
enum Role {
  USER   enum       // Enumerators user their name as value, unless explecitly set: "User"
  ADMIN  enum = "A" // Value explicitly set to "A"
}
```

---

```
meta = {
    from = mongo
    name = "people_profiles"
}
```

Should `from` be `source` instead? It does select a named source after all.

---

```
author   User(id)
```

The 'id' field on the `User` model is marked as `@primary`, so thee is no need to specify it for this relation:

```
author   User
```

Our VS Code plugin should have a hint to remove the redundancy

---

```
// nullable array items and nullable photos field
  photos   Photo?[]?
```

There is no case where the photos field will be null.
Photo does not have a back-relation field, so we are dealing with a 1:m relation. For SQL connectors, the `Photo` table will have a column to store a single `Profile` id.
If there are no `Photo` rows that relate to a given `Profile`, the `photos` field will return an empty array, not null.

---

```
// anonymous embed (optional)
  embed size {
    height  int
    width   int
  }?
```

I suggest we settle on `fieldName embed` as this is prefered by Johannes, Marcus and I:

```
// anonymous embed (optional)
  size embed {
    height  int
    width   int
  }?
```

---

```
// mixin Post's fields into MoviePost
  // Based on Go's struct embedding syntax
  Post
```

Let's not include this in the spec for now. Either have a section for furture consideration or remove it.

---

```
// Comments in this datamodel have meaning. Look to godoc for inspiration:
// https://blog.golang.org/godoc-documenting-go-code

// Comments directly above a model are attached to the model
model Category {
```

Documenting comments should also be supported for fields. We should strongly consider differentiating between comments and documentation like Rust does: https://facility9.com/2016/05/writing-documentation-in-rust/ 

# Feedback on Resolved Questions

```
Can we improve comma / multi-line support?
```

The answer does not actually show use of multiple lines. Please add this.

---

The only section without an answer is this:

```
Does `Model@field` make sense for relations?
```

We have settled on `Model(field)`. Please add an answer to this section :-)

---

# Feedback to TODO sections

```
TODO: connector or data source?
```

A connector is the implementation of a connector for a specific API or Database.
A data source is an instantiation of a connector

---

```
Decimal
The decimal type should map to a predefined configuration of the exact precision type. We have used NUMERIC(precision = 65, scale = 30) since the beginning of Graphcool and has never received a complaint, so we will use that as the default on SQL storage engines. It is possible to use type specification to change this.

Elastic seems uninterested in supporting Decimal (BigDecimal): https://github.com/elastic/elasticsearch/issues/17006

Mongo supports decimal (128-bit decimal floating point, not configurable)

TODO: should we leave this to type specifications?
```

I propose that we include Decimal and accept that it's actual implementation varies between databases. Type specification and constraints can be used to force each database to use a lowest common denominator if so desired.

---

```
TODO: Is there a technical reason for the underscore? Is it just to avoid name clashes?
```

This is just to avoid name clashes. We do not allow model names to start with `_`, but people are still free to set the dbName to for example `_BlogsWriters`, so this is not a full guarantee. But name clash is much more unlikely.

---

```
TODO: This will be a very common path will be to start with an implicit many-to-many and eventually transition to an explicit join table. We'll need to make this flow as smooth as possible. I'm worried people will forget subtle things like the unique(blog, author)
```

A nice way to do this is to provide IDE support for going in both directions.

---

```
TODO: try and map out check constraints, it's really weird in this case.

TODO: this feels like it could be combined with type specification.
```

I don't think constraints are weird at all. They are simple checks performed by Prisma before writing to the underlying database.
Sometime it makes sense to enforce a constraint both using a type specification and a constraint, but not always.

---

```
TODO: I don't have a good mental model of the difference between block configuration & block attributes. They feel like the same thing to me. Should we consolidate?
```

If we support block attributes, then I don't think we should also support arbitrary block configuration. Maybe we limit block configuration to always be in the meta block?

---

```
TODO: where does User persist in this case?
```

When there are multiple data sources in a datamodel, then each model declaration must explicitly define which one they belong to. It should be possible to designate one datasource as default.

---

```
TODO:
Should we use colons instead of equals for better consistency with the JSON primitive type?
```

I don't understand this question :-)

---

```
TODO: are cascading deletes a connector concern or something in core? For example, does Mongo support these?
```

My current understanding is that cascading deletes are a core Prisma primitive. All our existing connectors support it. Mongos is implemented at the Prisma layer, and we intend to do the same for other databases without native support.

---

```
TODO: I could see a potential use case for this, where if you delete the join table, the entire relationship goes away.
```

If you want this, you should create an explicit join model.

---

# Feedback on Open Questions

```
Should we have named arguments?
```

Unless there are strong concerns, then I suggest we adopt Marcus's proposal:

```
Directives with 1 argument may omit the argument name. E.g. @alias("my_weight") and @alias(name:"my_weight") is allowed. Directives with multiple arguments must always specify all argument names.
```

matthewmueller commented:
Decisions from Meeting(s)

- Commas, no commas:

 No commas, no `@@` syntax. Put attributes below again.

- Version Dependencies:

 We'll keep the different datasources (`mgo1`, `mgo2`)

```
source mgo1 {
  type = "mongo"
  url = env("MONGO_URL")
}

source mgo2 {
  type = "mongo"
  version = 2
  url = env("MONGO_URL")
}

// id is backed in mongo's ObjectID
type User {
  id String @primary @mgo1.ObjectID 
}
```

- `@default` or `=` ?

If we go back to `name String = "..."`. For formatting, not sure on if it's a 4th column or not, but we'll leave this decision to the implementor.

- Type Declarations



- Enum syntax?

Tentatively:

```ts
enum Color {
  Red = 0
  Teal = 1
}
```

Needs a final decision. Spend an hour mapping enums to languages and databases. What input do we accept and what output do we return from a Photon perspective?

- Relation Syntax?

Do we have special syntax for relations or use `@relation` attribute? 

Needs a final decision

- Extending the datamodel

We haven't really mapped out datamodels that persist across databases. 

Needs decision. One possibility:

```
source mgo {
  type = "mongo"
  url = env("MONGO_URL")
}

source pg {
  type = "postgres"
  url = env("POSTGRES_URL")
  default = true
}

source pg2 {
  type = "postgres"
  version = 2 
  url = env("POSTGRES_URL")
}

// id is backed in mongo
type mgo.User {
  id String @primary @mgo1.ObjectID 
}

// email is backed in postgres
type pg2.User {
  email  String  @pg2.Citext
}

// user model backed by default
// in this case pg
type User {
  mgo.User
  pg2.User

  first_name  String
}
```

```
config User {
  id  @postgres.UUID
}
```

```
@primary = @unique @serial @postgres.UUID
```

- Do we keep type aliases?

type Numeric Int @postgres.Numeric(10, 5)

Motivate more

- How powerful are generators?

If we have the following:

```groovy
type Plan {
  // using the postgres money
  amount  Int  @postgres.Money
}
```

A:

- Generator configuration to allow higher-level inputs and outputs from the generate clients.
- We don't want to directly depend on a specific Money library from JS / Go 

Q: How will we disambiguate relations? In v1.1 we used relation names for this. My guess is that we will stick to this. How will this look like?



- What are our Core attributes (default, unique, ...)?



What do we do for custom types that don't have a clear backing to a single type (e.g. `postgres.Point`)?

```groovy
model Customer {
  id       Int             @primary
  address  String
  point    postgres.Point
  point       Int             @postgres.Numeric(5)
  amount  String          @postgres.Money
}
```

Our current thinking is:

-  For custom types that have a clear backing are type specifications (e.g. `Int @postgres.Numeric(5)`. -  For types that are complex (e.g. Point that has an X and a Y value), Postgres exports namespaces you can use in the type field.

```
model Customer {
  id       Int             @primary
  address  String
  point      postgres.Point
  point       Int             @postgres.Numeric(5)
  amount  String          @postgres.Money
}
```

Pros:
- Clear client types

Downsides:
- subtle, when you should use type specs

Questions:
- what about typespecifications where you don't know how the DB maps it underneath? (e.g. @postgres.Money)

Alternatives:

```
model Customer {
  id       Int             @primary
  address  String
  point      Point
  point       Int             @postgres.Numeric(5)
  amount  String          @postgres.Money
}

embed Point {
  X Int
  Y Int
} @postgres.Point
```

@matthewmueller  will think more about this

sorenbs commented:
### Replacing PRISMA_CONFIG

> Prisma 1 documentation: https://www.prisma.io/docs/prisma-server/local-prisma-setup-je3i/

In Prisma 1.x PRISMA_CONFIG is used to configure Prisma on the infrastructure level. It has the following properties:

 - managementApiSecret
 - port
 - databases

Prisma 2 does not have a `managementApiSecret`. The `port` that Prisma uses to communicate is managed transparently by Prisma Client.

#### Databases

The new datamodel lifts the database configuration into the datamodel.

Prisma 1:

```yml
    environment:
      PRISMA_CONFIG: |
        managementApiSecret: my-server-secret-123
        port: 4466
        databases:
          default:
            connector: mysql
            host: mysql-db
            port: 3306
            user: root
            password: prisma
```

Prisma 2:

```groovy
source mysql {
  type = "mysql"
  url = "mysql://mysql-db:3306/mydb?user=root&password=prisma"
}
```

This declares a source called mysql of type mysql connecting to the mydb schema in the datase instance running on port 3306 on the host mysql-db.

Notice that we are standardising on urls for all database connection configuration. In the .NET world there is a standard called [connection strings](https://www.connectionstrings.com/), but I don't think this is used outside .NET, so we should probably not rely on it.

matthewmueller commented:
@sorenbs thanks for getting started on this! I think connection strings are fairly well supported outside the .NET community. For example:

- [node-postgres](https://node-postgres.com/features/connecting#connection-uri)
- [go-sql-driver](https://github.com/go-sql-driver/mysql#dsn-data-source-name)
- [go-postgres](https://github.com/lib/pq)

I guess Pear DB uses it as well.

I feel quite strongly that we should advocate the community towards this direction, since it's just a simpler model. Everything is a URL.

One point on username and password, I'd like to change it to:

```
mysql://root:prisma@mysql-db:3306/mydb?connectionLimit=99
```
